src/connect.rs

```diff
@@ -5,95 +5,225 @@ use tokio::{
 };
 
 use tungstenite::{
-    client::uri_mode, handshake::client::Response, protocol::WebSocketConfig, Error,
+    client::uri_mode,
+    error::{Error, UrlError},
+    handshake::client::{Request, Response},
+    protocol::WebSocketConfig,
 };
 
-use super::{client_async_with_config, IntoClientRequest, Request, WebSocketStream};
+use super::{client_async_with_config, IntoClientRequest, WebSocketStream};
 
-#[cfg(feature = "tls")]
 pub(crate) mod encryption {
-    use native_tls::TlsConnector as NativeTlsConnector;
-    use tokio_native_tls::{TlsConnector as TokioTlsConnector, TlsStream};
+    use std::{
+        marker::PhantomData,
+        pin::Pin,
+        task::{Context, Poll},
+    };
 
-    use tokio::io::{AsyncRead, AsyncWrite};
+    use pin_project::pin_project;
+    use tokio::io::{AsyncRead, AsyncWrite, ReadBuf};
 
-    use tungstenite::{stream::Mode, Error};
-
-    use crate::stream::Stream as StreamSwitcher;
+    use crate::stream::Stream;
 
     /// A stream that might be protected with TLS.
-    pub type MaybeTlsStream<S> = StreamSwitcher<S, TlsStream<S>>;
+    pub type MaybeTlsStream<S> = Stream<S, TlsStream<S>>;
+
+    #[allow(missing_docs)]
+    #[non_exhaustive]
+    #[pin_project(project = StreamProj)]
+    pub enum TlsStream<S> {
+        Plain(PhantomData<S>),
+        #[cfg(feature = "native-tls")]
+        NativeTls(tokio_native_tls::TlsStream<S>),
+        #[cfg(feature = "rustls-tls")]
+        Rustls(tokio_rustls::client::TlsStream<S>),
+    }
```

Comment:
Would it be possible to merge (functionality of) `TlsStream` and `crate::stream::Stream`? From a quick look, such a merge could prevent some logical duplication, and could avoid TlsStream::Plain being a dummy implementation doing nothing.

A natural name for such a merged type would be `MaybeTlsStream`. (i.e. it would take over the alias).

Reply:
Exactly my thought. 

I simply created new types as an experiment to see how well my idea works. In the end I saw that too but didn't have a chance to merged that with `crate::stream::Stream` yet.

Really like the idea of updating the name to `MaybeTlsStream`.